ecosystem/sep-0045.md

```diff
@@ -0,0 +1,443 @@
+Track: Standard
+Status: Draft
+Created: 2024-10-08
+Updated: 2024-10-08
+Version: 0.1.0
+Discussion: TBA
+```
+
+## Simple Summary
+
+This SEP defines the standard way for clients such as wallets or exchanges to create authenticated web sessions on
+behalf of a user who holds a Contract account. A wallet may want to authenticate with any web service which requires a
+contract account ownership verification, for example, to upload KYC information to an anchor in an authenticated way as
+described in [SEP-12](sep-0012.md).
+
+This SEP also supports authenticating users of shared contract accounts. Clients can use [memos](#memos) to distinguish
+users or sub-accounts of shared accounts.
+
+This SEP is based on [SEP-10](sep-0010.md), but only supports contract accounts and aims to provide a JWT that is fully
+compatible with existing protocols that use SEP-10.
+
+## Abstract
+
+This protocol is a variation of mutual challenge-response, which uses Soroban authorization entries to encode challenges
+and responses.
+
+It involves the following components:
+
+- A **Home Domain**: a domain hosting a [SEP-1 stellar.toml](sep-0001.md) containing a `WEB_AUTH_ENDPOINT_SEP0045` (URL)
+  and `SIGNING_KEY` (`G...`).
+- A **Server**: a server providing the `WEB_AUTH_ENDPOINT_SEP0045` that implements the GET and POST operations discussed
+  in this document. The server's domain may be the **Home Domain**, a sub-domain of the **Home Domain**, or a different
+  domain.
+  - The `SIGNING_KEY` from the **Home Domain** is the **Server Account**.
+- A **Client Account**: the account being authenticated.
+  - A Contract address (`C...`) that may be accompanied by a memo to scope the authentication to a user or sub-account
+    of the account.
+- A **Web Auth Contract**: a contract that implements the `web_auth_verify` function as described in this document. The
+  contract must be deployed at the `WEB_AUTH_CONTRACT_ID` address specified in the **Server**'s `stellar.toml`.
+
+The discovery flow is as follows:
+
+1. The **Client** retrieves the `stellar.toml` from the **Home Domain** in accordance with
+   [SEP-1 stellar.toml](sep-0001.md).
+1. The **Client** looks up the `WEB_AUTH_ENDPOINT_SEP0045`, `WEB_AUTH_CONTRACT_ID` and `SIGNING_KEY` (i.e. **Server
+   Account**) from the `stellar.toml`.
+
+The authentication flow is as follows:
+
+1. The **Client** requests a unique [`challenge`](#challenge) from the **Server** which includes a list of Soroban
+   authorization entries and the server's signatures represented as XDR strings.
+1. The **Server** responds with the challenge.
+1. The **Client** verifies that each authorization entry is signed by the **Server Account**.
+1. The **Client** verifies that each authorization entry does not contain any sub-invocations.
+1. The **Client** signs the first authorization entry using the secret key(s) of the signer(s) for the **Client
```

Comment:
Good point, we only need to verify `__check_auth` succeeds, which does not always require client signatures. The current API assumes that each authorization entry requires a signature, so that must be fixed before I rephrase this.